### PR TITLE
feat: Add ability to upload PT/TF models to Huggingface Hub

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR is a follow up on @fg-mindee  object detection `from_hub` and adds the following functionality:

- you can add `--push-to-hub` as flag to your training script 
- This does ask to login (if not already logged in)  to HF hub
- creates a repo on the hub for your model (same name as experiment)
- converts the model in the needed format / saves model configuration
- adds a README to the repo with some information
- and finally push to the hub :smiley: 

Any feedback is very welcome :hugs: 
@osanseviero Suggestions from your side as to whether we have forgotten anything or can do better ? :smile: 

@fg-mindee @charlesmindee 
I am not sure if it is the correct place for the file ..maybe add in doctr.utils.hub to add some tests?? 

Example Uploads:
https://huggingface.co/Felix92/Test-recognition-PyTorch
https://huggingface.co/Felix92/Test-recognition-TensorFlow

Note: PR include cfg fix for TF Resnet50 and FasterRCNN link to #883 (sry for the mistake 😅 )

### CODE REVIEW

**Comment by felixdittrich92** [comment]
Depends on: #883

File: references/hub.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+>>> img = DocumentFile.from_images(['<image_path>'])
+>>> # Load your model from the hub
+>>> model = from_hub('mindee/my-recognition-model').eval()
+>>> # Pass it to the predictor
+>>> predictor = ocr_predictor(det_arch='db_mobilenet_v3_large',
+>>>                           reco_arch=model,
+>>>                           pretrained=True)
+>>> # Get your predictions
+>>> res = predictor(img)
+```
```
Comment (charlesmindee) [inline references/hub.py:90]: I think it is not the right docstring there

**Review by charlesmindee — COMMENTED** [review]
Thanks for this amazing feature, for the location I think it might be better to include it in the package in a models/<task>/factory folder or in a models/factory folder as @fg-mindee did for the from_hub() method (since it is a symmetrical operation and can be decorrelated from training).
What do you think ?

File: references/hub.py (hunk 1 above)
Comment (felixdittrich92) [reply references/hub.py:90]: @charlesmindee 
Yes i think we have to change it ftm to TODO or anything else until classifiaction/detection/recognition have a `from_hub`

I'm fine with a change to models/factory but i am not sure how we can test it in fact that we would need a huggingface test account for this 

Example from model card currently
Screenshot from 2022-04-04 11-45-14

Comment (felixdittrich92) [reply references/hub.py]: @charlesmindee 

sry i think you misunderstood :) this is a readme file which will be added to the uploaded model-repository but now i have added also docstrings :+1:

File: doctr/models/factory/hub.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    # check if git lfs is installed
+    try:
+        subprocess.call(['git', 'lfs', 'version'])
+    except FileNotFoundError:
+        raise OSError('Looks like you do not have git-lfs installed, please install. \
+                      You can install from https://git-lfs.github.com/. \
+                      Then run `git lfs install` (you only have to do this once).')
+
+
+def _save_model_and_config_for_hf_hub(model: Any, save_dir: str, model_config: Dict[str, str]) -> None:
```
Comment (frgfm) [inline doctr/models/factory/hub.py:41]: Model config typing in incorrect (values can be non string)

**Review by frgfm — COMMENTED** [review]
Thanks for the PR Felix 🙏

Regarding the general idea of this feature, it needs to be complementary to "from_hub". For this, I would suggest that we try to find a way to imbue the config structure by default in the function. (Perhaps passing the content as args or kzwargs and creating the dictionary in the function)

Either way, this function needs to have a reduced minimal snippet to push a model in a shell.

So I'd argue that we need to ensure that :
- we create models with a comprehensive config attribute 
- we use that attribute to create the config and push to hub 
- this config is then used with from_hub to recreate the model 


About testing, let's narrow down what we want to test:
- the HF api is well tested, so we need to focus on our own lines of code 
- the process here is to create a dictionary, dump parameters, create a git repo with it and push it
- with temporary dirs of pytest, I'd argue we can go up until the creation of the repo 
- the question now is how easily we can split those steps 

What do you think? 🙂

File: doctr/models/factory/hub.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    if is_torch_available():
+        import torch
+        weights_path = save_directory / 'pytorch_model.bin'
+        torch.save(model.state_dict(), weights_path)
+    elif is_tf_available():
+        import tensorflow as tf  # noqa: F401
+        weights_path = save_directory / 'tf_model' / 'weights'
+        model.save_weights(str(weights_path))
+    else:
+        raise RuntimeError("Neither PyTorch nor TensorFlow is available.")
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/models/factory/hub.py:61]: Sure 😅

File: doctr/models/factory/hub.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+### Example usage:
+
+```python
+>>> from doctr.io import DocumentFile
+>>> from doctr.models import ocr_predictor
+>>> from doctr.models.<task> import from_hub
+>>>
+>>> img = DocumentFile.from_images(['<image_path>'])
+>>> # Load your model from the hub
+>>> model = from_hub('mindee/my-recognition-model').eval()
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/models/factory/hub.py:104]: 👍

File: doctr/models/factory/hub.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+{json.dumps(vars(run_config))}
+"""
+
+    # add model configuration
+    model_config = model.cfg
+    model_config['arch'] = run_config.arch
+
+    commit_message = f'Add {model_name} model'
+
+    local_cache_dir = os.path.join(os.path.expanduser('~'), '.cache', 'doctr', 'huggingface_cache', model_name)
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/models/factory/hub.py:123]: 👍

File: doctr/models/factory/hub.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+__all__ = ['login_to_hub', 'push_to_hf_hub']
+
+
+def login_to_hub() -> None:
+    """Login to huggingface hub
+    """
+
+    access_token = HfFolder.get_token()
+    if access_token is not None and HfApi()._is_valid_token(access_token):
+        print("Huggingface Hub token found and valid")
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/models/factory/hub.py:27]: Yes sure forgot after moving away from references :)

File: doctr/models/factory/hub.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    """
+
+    save_directory = Path(save_dir)
+
+    if is_torch_available():
+        import torch
+        weights_path = save_directory / 'pytorch_model.bin'
+        torch.save(model.state_dict(), weights_path)
+    elif is_tf_available():
+        import tensorflow as tf  # noqa: F401
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply doctr/models/factory/hub.py:57]: 👍

File: doctr/models/factory/hub.py (hunk 2 above)
Comment (felixdittrich92) [reply doctr/models/factory/hub.py:41]: 👍

**Comment by felixdittrich92** [comment]
@frgfm 
Nice that you are back :hugs: :+1: 

For the config part:
- i use currently the `model.cfg` (#883 fixes the last) this are updated while training otherwise its the `default_cfg` i think this is enough to rebuild the model additional there is stored the model arch and now also the task (recognition/detection/classification/object_detection)

- i have added some tests i think that will cover most of our site. wdyt ?

**Comment by felixdittrich92** [comment]
@frgfm 
wdyt ? :hugs:

**Comment by fharper** [comment]
@charlesmindee can you follow up on the discussion FG started for this PR please?

**Comment by felixdittrich92** [comment]
@charlesmindee failing test has nothing to do with this PR please trigger again if we are fine with the other stuff :)

**Review by charlesmindee — APPROVED** [review]
Thanks for the tests & the edits!

**Comment by fharper** [comment]
Thanks for the PR @felixdittrich92 🎉

**Comment by frgfm** [comment]
Missing "type: new feature", "module: models", "ext: references" and "ext: test" as PR labels I think @charlesmindee :)
